File: MyLeetCode_Go/study/LeetCode/Problem_575_DistributeCandies.go
package main
//
//func distributeCandies(candyType []int) int {
//	if candyType == nil || len(candyType) == 0 {
//		return 0
//	}
//
//	// 利用 map 的 key 唯一性当 set 使用
//	//set := map[int]bool{}
//	candyMap := map[int]int{}
//	for i:=0;i< len(candyType);i++ {
//		if _, ok := candyMap[i]; !ok {
//			candyMap[i] = 1
//		} else {
//			candyMap[i]++
//		}
//	}
//
//	return process(candyType, 0, candyMap, len(candyType)/2)
//}
//
//
//// 当前来到 i 位置分配, 之前0~i-1已经分配完了
//// rest: 待分给妹妹剩余糖果数
//// 返回: 最大种类
//func process(candyType []int, index int, candyMap map[int]int, rest int) int {
//	if index == len(candyType) {
//		return 0
//	}
//	// index < len
//
//	// 当前糖果分给妹妹
//	p1 := -1
//	nextA := process(candyType, index+1, )
//
//}

File: MyLeetCode_Java/src/main/java/LeetCode/Problem_575_DistributeCandies.java
package LeetCode;

import java.util.*;

public class Problem_575_DistributeCandies {

    private int maxKind = 0;

    // 暴力解法, 每一个糖果选择去 A,或 B,最后去重
    public int distributeCandies1(int[] candyType) {
        if (candyType == null || candyType.length == 0) {
            return 0;
        }
        List<Integer> path = new ArrayList<>();
        process(candyType, 0, candyType.length / 2, path);
        return maxKind;
    }

    // 当前来到 i 位置, 剩余待处理 rest
    // 收集所有结果
    private void process(int[] candyType, int i, int rest, List<Integer> path) {
        if (i == candyType.length) {
            // 统计 path 种类
            maxKind = Math.max(maxKind, getKind(path));
            return;
        }
        // i 位置给 A
        if (rest > 0) {
            path.add(candyType[i]);
            process(candyType, i + 1, rest - 1, path);
            path.remove(Integer.valueOf(candyType[i]));
        }
        // i位置不给 A
        process(candyType, i + 1, rest, path);
    }

    private int getKind(List<Integer> path) {
        Set<Integer> set = new HashSet<>(path);
        return set.size();
    }

    // 贪心
    public int distributeCandies2(int[] candyType) {
        if (candyType == null || candyType.length == 0) {
            return 0;
        }

        int N = candyType.length;
        Set<Integer> set = new HashSet<>();
        for (int candy: candyType) {
            set.add(candy);
        }
        return Math.min(set.size(), N / 2);
    }

    public static void main(String[] args) {
        Problem_575_DistributeCandies sl = new Problem_575_DistributeCandies();

//        int[] candies = new int[]{1,1,2,2,3,3};
        int[] candies = new int[]{3,3,3,3};
//        int[] candies = new int[]{1, 1, 2, 3};

        int ans = sl.distributeCandies1(candies);
        System.out.println(ans);

    }
}
